tests/cases/compiler/contextualTypeCaching.ts

```diff
@@ -0,0 +1,25 @@
+// @strict: true
+// @noEmit: true
+
+// Repro from #52575
+
+export interface Event<T> {
+    callback: (response: T) => void;
+    nested: {
+        nestedCallback: (response: T) => void;
+    }
+}
+
+export type CustomEvents = {
+    a: Event<string>
+    b: Event<number>
+};
+
+declare function emit<T extends keyof CustomEvents>(type: T, data: CustomEvents[T]): void
+
+emit('a', {
+    callback: (r) => {},
+    nested: {
+        nestedCallback: (r) => {},
+    },
+});

```

Comment:
This is a weird one. It currently works OK on the `main` branch - so it isn't fixed by this PR per se. However, one of my PRs regressed this case - it was caught by the bot [here](https://github.com/microsoft/TypeScript/pull/52589#issuecomment-1416180638). I've re-checked that your PR here fixes it (see [here](https://github.com/microsoft/TypeScript/pull/52589#issuecomment-1416812513)) and that this "regression" was introduced with the `const` type parameters PR. Somehow my PR impacted things for this case to only manifest with it 🤷‍♂️ 

I figured out that it might be worthwhile to include this test case here even if it isn't exactly fixed by this PR **right now**. 

```suggestion
});

// repro from 52589#issuecomment-1416180638
declare class MyCompiler {
  compile(): void;
}

interface WebpackPluginInstance {
  apply: (compiler: MyCompiler) => void;
}

type WebpackPluginFunction = (this: MyCompiler, compiler: MyCompiler) => void;

interface Optimization {
  minimizer?: (WebpackPluginInstance | WebpackPluginFunction)[];
}

declare const A: <T, P extends keyof T>(
  obj: T,
  prop: P,
  factory: () => T[P]
) => void;

const applyOptimizationDefaults = (optimization: Optimization) => {
  A(optimization, "minimizer", () => [
    {
      apply: (compiler) => {},
    },
  ]);
};

```

Reply:
Would be nice to get more code coverage in these scenarios.